### PR TITLE
lang/rust: Migrate to avro-rs

### PR DESCRIPTION
**Pull request by Xuanwo**
## What is the purpose of the change

This PR migrates the rust codebase to [avro-rs](https://github.com/apache/avro-rs)

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -452,11 +452,7 @@
                 <exclude>lang/ruby/.gem/**</exclude>
                 <exclude>lang/ruby/pkg/**</exclude>
                 <exclude>lang/ruby/.bundle/**</exclude>
-                <exclude>lang/rust/target/**</exclude>
-                <exclude>lang/rust/precommit_venv/**</exclude>
-                <exclude>lang/rust/Cargo.lock</exclude>
-                <exclude>lang/rust/README.tpl</exclude>
-                <exclude>lang/rust/.requirements-precommit.txt</exclude>
+                <exclude>lang/rust/README.md</exclude>
```
Comment (martin-g) [inline pom.xml:455]: @Xuanwo What do you use instead of RAT for license checking in your projects ?

File: lang/rust/README.md [hunk 2]
```diff
@@ -1,46 +1,5 @@
...
-See [avro_derive/README.md](./avro_derive/README.md)
-
-## Avro test helper
-
-See [avro_test_helper/README.md](./avro_test_helper/README.md)
-
-## WebAssembly demo application
-
-See [wasm-demo/README.md](./wasm-demo/README.md)
+We apologize for the inconvenience and greatly appreciate your contributions!
```
Comment (martin-g) [inline lang/rust/README.md:5]: ❤️

**Review by martin-g — APPROVED** [review]

**Comment by martin-g** [comment]
```
pypy3.9: install_deps> python -I -m pip install coverage python-snappy zstandard
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [6 lines of output]
      Checking for Rust toolchain....
      
      Cargo, the Rust package manager, is not installed or is not on PATH.
      This package requires Rust and Cargo to compile extensions. Install it through
      the system's package manager or via https://rustup.rs/
      
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed
```
It seems the Python tests depend on the Rust toolchain.

File: lang/rust/README.md (hunk 2 above)
Comment (Gordon-F) [reply lang/rust/README.md:5]: just curious, why only rust version moved to its own repo?

File: pom.xml (hunk 1 above)
Comment (Xuanwo) [reply pom.xml:455]: Hi, I'm using https://github.com/korandoru/hawkeye which written in rust and provides github actions natively:

```yml
licenses:
  runs-on: ubuntu-latest
  timeout-minutes: 10
  env:
    FORCE_COLOR: 1
  steps:
    - uses: actions/checkout@v4
    - name: Check license headers
      uses: korandoru/hawkeye@v5
```

**Comment by Xuanwo** [comment]
> Linking to the dev mailing list thread where the rationale was explained: [[VOTE] Extract the Rust SDK into its own Git repository](https://lists.apache.org/thread/rpvgxrgsc7obv3qnj5zsmnkp4z112g0g)

Thank you for the suggestion; it's been added to the Rust README now.

**Review by martin-g — APPROVED** [review]

**Comment by martin-g** [comment]
Thank you, @Xuanwo !

**Comment by Xuanwo** [comment]
> Some python tools are written in rust:
> 
> ```
>         --- stderr
>         configure.ac:22: installing 'build-aux/compile'
>         configure.ac:24: installing 'build-aux/config.guess'
>         configure.ac:24: installing 'build-aux/config.sub'
>         configure.ac:11: installing 'build-aux/install-sh'
>         configure.ac:11: installing 'build-aux/missing'
>         Makefile.am: installing 'build-aux/depcomp'
>         parallel-tests: installing 'build-aux/test-driver'
>         configure: error: No modern nasm or yasm found as required. Nasm should be v2.11.01 or later (v2.13 for AVX512) and yasm should be 1.2.0 or later.
>         thread 'main' panicked at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/isal-sys-0.3.1+496255c/build.rs:72:17:
>         configure failed
>         note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
>       warning: build failed, waiting for other jobs to finish...
>       💥 maturin failed
>         Caused by: Failed to build a native library through cargo
>         Caused by: Cargo build finished with "exit status: 101": `env -u CARGO PYO3_ENVIRONMENT_SIGNATURE="pypy-3.9-64bit" PYO3_PYTHON="/home/runner/work/avro/avro/lang/py/.tox/py/bin/python" PYTHON_SYS_EXECUTABLE="/home/runner/work/avro/avro/lang/py/.tox/py/bin/python" "cargo" "rustc" "--message-format" "json-render-diagnostics" "--manifest-path" "/tmp/pip-install-qqb1t7fw/cramjam_87d55c093e714efabcee81485a095f12/Cargo.toml" "--release" "--lib" "--" "-C" "strip=symbols"`
>       Error: command ['maturin', 'pep517', 'build-wheel', '-i', '/home/runner/work/avro/avro/lang/py/.tox/py/bin/python', '--compatibility', 'off'] returned non-zero exit status 1
>       [end of output]
> ```
> 
> I'm working on this now.

Oh, they have been failing for a long time, even before this PR.

File: lang/rust/README.md (hunk 2 above)
Comment (Gerrit0) [reply lang/rust/README.md]: It's an experiment, if it ends up being better for sustained development, the other languages will likely be split out as well.

**Pull request merged**